### ISSUE TITLE
Questions about phpunit critical security vulnerability

### ISSUE THREAD

**Issue by ghost** [PrestaShop/PrestaShop#17059]
https://build.prestashop.com/news/critical-security-vulnerability-in-prestashop-modules/

Please clarify - you say root /vendor folder. Do you mean also the Symfony related folder /vendor/symfony/symfony/src/Symfony/Bridge/PhpUnit/ too?

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/prestashop-core-developers

**Comment by matks** [comment]
Hi @rss9090, the folder /vendor/symfony/symfony/src/Symfony/Bridge/PhpUnit/ is safe.

The directories which contain the vulnerability are the original [PHPUnit](https://github.com/sebastianbergmann/phpunit) directories that can only be found in
- <prestashop_directory>/vendor
- <prestashop_directory>/modules/<module_name>/vendor

And, to be more accurate, the vulnerability in these vendors is in file [src/Util/PHP/eval-stdin.php](https://github.com/sebastianbergmann/phpunit/blob/master/src/Util/PHP/eval-stdin.php).

**Comment by adriansuter** [comment]
Would it make sense to check the server logs for requests to this `eval-stdin.php` script?

If so, maybe that could be added as additional information to the page https://build.prestashop.com/news/critical-security-vulnerability-in-prestashop-modules/

I have seen in the last week in my server logs (not even prestashop sites) a lot of requests against `/vendor/phpunit/phpunit/phpunit.xml`. Maybe that are already attempts.

**Comment by PierreRambaud** [comment]
This kind of check must be done with your system administrator. 
Logs are not managed all in the same way, and having a huge logs could increase the execution time of your code.
If you need more information about how you can manage your logs easily have a look at https://www.elastic.co/what-is/elk-stack

**Comment by ghost** [comment]
Hi @matks thank you for your answer.

I see in /modules/ps_facetedsearch/vendor/bin/phpunit file. 
Should it been deleted also or it considered as safe?

You said that  folder /vendor/symfony/symfony/src/Symfony/Bridge/PhpUnit/  is safe but whether the Prestashop use it or it can be deleted with no problems?

> And, to be more accurate, the vulnerability in these vendors is in file src/Util/PHP/eval-stdin.php. 

no eval-stdin.php file there in my case

**Comment by matks** [comment]
> I see in /modules/ps_facetedsearch/vendor/bin/phpunit file.
> Should it been deleted also or it considered as safe?

This file is safe but you can also delete it as it is useless for your shop. It is a file used to run tests, which is never going to happen on your shop.

> You said that folder /vendor/symfony/symfony/src/Symfony/Bridge/PhpUnit/ is safe but whether the Prestashop use it or it can be deleted with no problems?

I think you can delete it if you want 🤔but I cant guarantee 100% . These files are helper files used by Symfony to run tests powered by PHPUnit. As far as we know, they are harmless.

> no eval-stdin.php file there in my case

eval-stdin.php is the file responsible for the issue. As long as you do not have it on your server, you're safe.

Actually we could tell people "instead of deleting vendor/phpunit file you can simply remove eval-stdin.php file" but it would be a little more complex to explain and since the vendor/phpunit file is useless anyway, deleting it is a good thing.

**Comment by parpotosis** [comment]
Which versions of PHPUNIT are vulnerable?
According to this https://nvd.nist.gov/vuln/detail/CVE-2017-9841 versions starting 5.6.3 could be safe.
But in this article https://build.prestashop.com/news/critical-security-vulnerability-in-prestashop-modules/ you write that "the issue is fixed in PHPUnit 7.5.19 and 8.5.1" [...] and "all previous versions are vulnerable, at least for certain server configurations."

**Comment by matks** [comment]
We have unfortunately confirmed that phpunit versions > 5.6.3 are vulnerable as well.

There is indeed the vulnerability https://nvd.nist.gov/vuln/detail/CVE-2017-9841 and there is **another vulnerability** that is not reported yet which impacts phpunit up to 7.5.19 | 8.5.1 (these 2 versions are the 2 patched versions).

This 2nd vulnerability has been reported to phpunit maintainer and is currently being reported to CVE authority in order to make register it officially.

**Comment by parpotosis** [comment]
Thanks


> We have unfortunately confirmed that phpunit versions > 5.6.3 are vulnerable as well.
> 
> There is indeed the vulnerability https://nvd.nist.gov/vuln/detail/CVE-2017-9841 and there is **another vulnerability** that is not reported yet which impacts phpunit up to 7.5.19 | 8.5.1 (these 2 versions are the 2 patched versions).
> 
> This 2nd vulnerability has been reported to phpunit maintainer and is currently being reported to CVE authority in order to make register it officially.

Thanks!

**Comment by adriansuter** [comment]
As a sidenote: PHPUnit should never be deployed to the web server. It is intended to be used for unit testing during development only.

**Comment by PierreRambaud** [comment]
@adriansuter PHPUnit could also be used for integration / production tests. Like Tempest do with Openstack (already done it with python / ruby tools).
Even if you don't have the phpunit directory in your public directory, the issue is still here and can be exploited with and Local file inclusion

**Comment by novaclic** [comment]
> We have unfortunately confirmed that phpunit versions > 5.6.3 are vulnerable as well.
> 
> There is indeed the vulnerability https://nvd.nist.gov/vuln/detail/CVE-2017-9841 and there is **another vulnerability** that is not reported yet which impacts phpunit up to 7.5.19 | 8.5.1 (these 2 versions are the 2 patched versions).
> 
> This 2nd vulnerability has been reported to phpunit maintainer and is currently being reported to CVE authority in order to make register it officially.

I understand the non disclosure at this stage. Could you please just let us know if the additional vulnerability affects the same file (`eval-stdin.php`) or not ?
(currently, asserting an intrusion is done by searching the Apache Log for a HTTP-POST on this file, it would be good to know if other files should be included).

Thanks !

**Comment by PierreRambaud** [comment]
@novaclic Only `eval-stdin.php` is compromised, the vulnerability is the same, it depends if you are running under CGI/FastCGI context which interpret `php://stdin` like `php://input`.

**Comment by KTN1990** [comment]
i found the bug on november 2019 and i founded that + 4k sites are vuln if prestashop had a bug bounty system its will be so helpful and too many people can report bugs

**Comment by PierreRambaud** [comment]
@KTN1990 This bug exists since many years, every websites having the phpunit on the server is currently vulnerable (not only PrestaShop instances). 
We have a security@prestashop.com e-mail as mentioned [here](https://github.com/PrestaShop/PrestaShop/security/policy). The bug bounty program is something we are discussing, but it costs time and money :wink: 
We received since years e-mails about security issues, so the bug bounty program isn't mandatory to report something, next time, feel free to help us if you found something strange on the Internet :wink:

**Comment by KTN1990** [comment]
@PierreRambaud i know about the phpunit bug its about 3 years or 4 years ago,  i understand that bug bounty costs time and money... but too many people from the underground love to post bugs but they need money and support thats why bug bounty can help any way thanks for the discussion.

**Comment by pdsolutions** [comment]
Hello,

DO NOT DELETE phpunit  directory without a full backup. Today i found and delete the phpunit directory, as described here: https://www.prestashop.com/en/security-announcement-your-store-vulnerable-malware and it broken my shop.

**Comment by PierreRambaud** [comment]
@pdsolutions Which phpunit directory do you delete?

**Comment by khouloudbelguith** [comment]
Hi @rss9090,

Since we had no news from you for more than 30 days, I'll close this ticket. Feel free to open another one if you can give specific details.

Thanks!

**Comment by zapalm** [comment]
The deletion of `phpunit` folders is not just wrong solution, it can also stop the site from working in some cases. Only one file is _potentially_ vulnerable and must be deleted: `eval-stdin.php`.

For ease of solving the problem, use my free module: [Tool for maintenance & debug](https://www.prestashop.com/forums/topic/432277-module-tool-for-maintenance-debug/). It seems that I did a better solution in this module to the problem than others, because the list of potentially infected files is also checked and the PHPUnit directories are not searched for and deleted. I have disclosed some details of the vulnerability in [this article](https://www.patreon.com/posts/58131075).
